dash/include/dash/halo/Halo.h

```diff
@@ -11,55 +11,150 @@
+    int16_t                               value, Values... values)
+  : Base_t::Dimensional(value, (int16_t) values...), _coefficient(coefficient) {
+  }
 
   // TODO as constexpr
+  /**
+   * Returns maximum distance to center over all dimensions
+   */
   int max() const {
     int16_t max = 0;
     for(dim_t i(0); i < NumDimensions; ++i)
       max = std::max((int) max, (int) std::abs(this->_values[i]));
     return max;
   }
-};  // Stencil
 
-template <dim_t NumDimensions, std::size_t NumStencilPoints>
+  /**
+   * Returns the coefficient for this stencil point
+   */
+  CoeffT coefficient() const { return _coefficient; }
+
+private:
+  CoeffT _coefficient = 1.0;
+};  // StencilPoint
+
+/**
+ * A collection of stencil points (\ref Stencil)
+ * e.g. StencilSpec<dash::StencilPoint<2>, 2,2>({StencilPoint<2>(-1,0),
+ * StencilPoint<2>(1,0)}) -> north and south
+ */
+template <typename StencilPointT, dim_t NumDimensions,
+          std::size_t NumStencilPoints>
 class StencilSpec {
 private:
-  using Self_t = StencilSpec<NumDimensions, NumStencilPoints>;
+  using Self_t = StencilSpec<StencilPointT, NumDimensions, NumStencilPoints>;
 
 public:
-  using stencil_size_t = std::size_t;
-  using Stencil_t      = Stencil<NumDimensions>;
-  using Specs_t        = std::array<Stencil_t, NumStencilPoints>;
+  using stencil_size_t  = std::size_t;
+  using stencil_index_t = std::size_t;
+  using Specs_t         = std::array<StencilPointT, NumStencilPoints>;
 
 public:
+  /**
+   * Constructor
+   *
+   * Takes a list of \ref StencilPoint
+   */
   constexpr StencilSpec(const Specs_t& specs) : _specs(specs) {}
 
+  /**
+   * Constructor
+   *
+   * Takes all given \ref StencilPoint. The number of arguments has to be the
+   * same as the given number of stencil points via the template argument.
+   */
+  template <typename... Values>
+  constexpr StencilSpec(const StencilPointT& value, const Values&... values)
+  : _specs{ { value, (StencilPointT) values... } } {
+    static_assert(sizeof...(values) == NumStencilPoints - 1,
+                  "Invalid number of stencil point arguments");
+  }
+
   // TODO constexpr
+  /**
+   * Copy Constructor
+   */
   StencilSpec(const Self_t& other) { _specs = other._specs; }
 
+  /**
+   * \return container storing all stencil points
+   */
   constexpr const Specs_t& specs() const { return _specs; }
 
+  /**
+   * \return number of stencil points
+   */
   static constexpr stencil_size_t num_stencil_points() {
     return NumStencilPoints;
   }
 
-  constexpr const Stencil_t& operator[](std::size_t index) const {
+  /**
+   * Finds the stencil point index for a given \ref StencilPoint
+   *
+   * \return The index and true if the given stecil point was found,
+   *         else the index 0 and false.
+   *         Keep in mind that index 0 is only a valid index, if the returned
+   *         bool is true
+   */
+  const std::pair<stencil_index_t, bool> index(StencilPointT stencil) const {
```

Comment:
`std::optional` is calling :)

Reply:
when we use c++17 as default, i will change it :)